### PR TITLE
Redux hot rewrite

### PR DESCRIPTION
**Pull request by jonsherrard**
Hi @shortlist-digital/owners 

I've rewritten all of the front end code for the plugin into, better-practice ES6, and using a state container library called `redux`

The app has a centralised state, so know you can navigate back and forth through stages, without losing your form entry so far. Eventually I'd like persist this in local storage too.

The app is setup in screens. It should be easy to build new screens such as a social login, and drop them in.

There's a nicer competitions select form.

Validations are done as you type with better visual feedback.

The T&Cs open in a modal, as do any errors we might get back from calais.

All in all it's a lot nicer I hope.

### CODE REVIEW

**Comment by iamdeaneyelid** [comment]
Very Trés!

**Comment by garethfoote** [comment]
This all looks good to me. Slightly beyond my current understanding of the 'redux' library but syntax-wise looks great. It appears to be influenced by ClojureScript + Elm. i.e. functional programming and immutable objects approach. Need to do some more reading in this area. 

:+1:

File: src/app.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+          {...dispatchers}
+          isStoreValid={() => false}
+        />
+      )
+    case THANK_YOU_SCREEN:
+      return <ThankYouScreen />
+
+    default:
+      return (
+        <h1 style={{textAlign:'center'}}>Something went dreadfully wrong</h1>
```
Comment (ecoad) [inline src/app.js:83]: :+1:

File: src/screens/thank-you.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import React, { Component } from 'react'
+
+class ThankYouScreen extends Component {
+  render() {
+    return <h1 style={{textAlign:'center'}}>We've submitted your entry succesfully!</h1>
```
Comment (ecoad) [inline src/screens/thank-you.js:5]: Has github rightfully picked up something incorrect here?

Comment (jonsherrard) [inline src/screens/thank-you.js:5]: Nope, the syntax highlighter doesn't have support for JSX in ES6 yet, and that unescaped ' breaks it.

File: src/state-builder.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
+// This is a bit mental to understand
```
Comment (ecoad) [inline src/state-builder.js:1]: Thanks for the commenting

**Comment by ecoad** [comment]
That's a great looking JS architecture, my hasn't she blossomed. I'm looking forward to more involvement in that React world. It strikes me that React Native apps will not be too much of a stretch of the imagination from here - to what extent would this be refactorable to a mobile app? (If you were posessed to do so)

**Comment by jonsherrard** [comment]
Yeah it's all very cross-compatible. The idea is that you actually refactor everything to the point where:

```
import SmartComponent from './smart-component'
import DumbComponent from './dumb-component'
<SmartComponent {...globalStore}> //logic
  <DumbComponent> //Just props and a render function using the DOM
</SmartComponent>
```

Then in your react-native main .js

```
import SmartComponent from './smart-component'
import DumbComponent from './dumb-component.ios.js' // or .android.js
<SmartComponent {...globalStore}> //logic
  <DumbComponent> //Just props and a render function using React Native UI interface
</SmartComponent>
```

Pretty crazy really!

**Comment by jonsherrard** [comment]
So when @garethfoote starts writing React / Redux models to interface with Parse.com, it's essentially writing large chunks of any hypothetical native map app too. :+1:

**Pull request merged**